cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -1203,6 +1203,48 @@ inline Status ConvertFixedSizeBinary(const ChunkedArray& data, PyObject** out_va
   return Status::OK();
 }
 
+inline Status ConvertStruct(const ChunkedArray& data, PyObject** out_values) {
+  PyAcquireGIL lock;
+  std::vector<OwnedRef> fields_data;
+  for (int c = 0; c < data.num_chunks(); c++) {
+    auto arr = static_cast<StructArray*>(data.chunk(c).get());
+    fields_data.resize(arr->fields().size());
+    // Convert the struct arrays first
+    for (size_t i = 0; i < arr->fields().size(); i++) {
+      PyObject* numpy_array;
+      RETURN_NOT_OK(ConvertArrayToPandas(arr->field(i), nullptr, &numpy_array));
+      fields_data[i].reset(numpy_array);
+    }
+
+    // Construct a dictionary for each row
+    const bool has_nulls = data.null_count() > 0;
+    for (int64_t i = 0; i < arr->length(); ++i) {
+      if (has_nulls && arr->IsNull(i)) {
+        Py_INCREF(Py_None);
+        *out_values = Py_None;
+      } else {
+        PyObject* item = PyDict_New();
+        RETURN_IF_PYERROR();
+        for (uint field_idx = 0; field_idx < arr->fields().size(); field_idx++) {
+          if (!arr->field(field_idx)->IsNull(i)) {
+            auto name = arr->type()->child(field_idx)->name();
+            auto array = reinterpret_cast<PyArrayObject*>(fields_data[field_idx].obj());
+            auto ptr = reinterpret_cast<const char*>(PyArray_GETPTR1(array, i));
```

Comment:
Yes, it cannot be null according to the current implementation of `ConvertArrayToPandas`.
We can still add a `DCHECK` to be safe in debug if we want be pedantic and protect against implementation change (probably better right after `ConvertArrayToPandas`)